packages/account/test/optimistic-concurrency-handling.test.ts

```diff
@@ -0,0 +1,440 @@
+import { ErrorCode, FuelError } from '@fuel-ts/errors';
+import { expectToThrowFuelError } from '@fuel-ts/errors/test-utils';
+import { sleep } from '@fuel-ts/utils';
+
+import { Provider } from '../src/providers';
+import { setupTestProviderAndWallets } from '../src/test-utils';
+
+/**
+ * @group node
+ * @group browser
+ */
+describe('optimistic concurrency handling via block height', () => {
+  afterEach(() => {
+    vi.restoreAllMocks();
+  });
+
+  test(`operations that aren't block-sensitive don't include the required block height in request`, async () => {
+    using launched = await setupTestProviderAndWallets();
+
+    const { provider } = launched;
+
+    const fetchSpy = vi.spyOn(global, 'fetch');
+
+    await provider.operations.getChain();
+
+    const body = JSON.parse(fetchSpy.mock.calls[0][1]?.body?.toString() as string);
+
+    expect(body.extensions?.required_fuel_block_height).toBeUndefined();
+  });
+
+  test(`waits when current block height is higher than actual and within node's tolerance`, async () => {
+    using launched = await setupTestProviderAndWallets({
+      nodeOptions: {
+        args: [
+          '--poa-instant',
+          'false',
+          '--poa-interval-period',
+          '50ms',
+          '--graphql-required-block-height-tolerance',
+          '10', // this is the default value but we're setting it explicitly for clarity
+        ],
+      },
+    });
+
+    const {
+      provider: launchedProvider,
+      wallets: [wallet],
+    } = launched;
+
+    const baseAssetId = await launchedProvider.getBaseAssetId();
+
+    Provider.clearChainAndNodeCaches();
+
+    const provider = new Provider(launchedProvider.url);
+
+    const {
+      chain: {
+        latestBlock: { height },
+      },
+    } = await provider.operations.getLatestBlockHeight();
+
+    const expectedHeight = +height + 8;
+
+    const fetchSpy = vi.spyOn(global, 'fetch').mockImplementationOnce((url, request) => {
+      // we're mocking it to modify the required_fuel_block_height
+      // but we still want the real fetch to be called
+      // hence why we reset immediately and return the real fetch call in the end
+      fetchSpy.mockReset();
+
+      const body = JSON.parse(request?.body?.toString() as string);
+      body.extensions.required_fuel_block_height = expectedHeight;
+
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      request!.body = JSON.stringify(body);
+
+      return fetch(url, request);
+    });
```

Comment:
For this test suite, would it be better, instead of mocking the returned value by the request, to ensure that the GraphQL request genuinely took longer than usual because it was waiting for the block to sync?

Maybe we could use:
```
--poa-interval-period 200ms
--graphql-required-block-height-tolerance 5
```

to simplify the validation.

Reply:
That's exactly what the mock is doing. I have reset the implementation on line `68`  so that the _actual_ fetch implementation is called and the node is hit.